cpp/src/copying/contiguous_split.cu

```diff
@@ -571,46 +630,55 @@ BufInfo build_output_columns(InputIter begin,
 {
   auto current_info = info_begin;
-  std::transform(begin, end, out_begin, [&current_info, base_ptr](column_view const& src) {
-    // Use C++17 structured bindings
-    bitmask_type const* bitmask_ptr;
-    size_type null_count;
-    std::tie(bitmask_ptr, null_count) = [&]() {
-      if (src.nullable()) {
-        auto const ptr = reinterpret_cast<bitmask_type const*>(base_ptr + current_info->dst_offset);
-        ++current_info;
-        return std::make_pair(ptr, UNKNOWN_NULL_COUNT);
-      }
-      return std::make_pair(static_cast<bitmask_type const*>(nullptr), 0);
-    }();
-    uint8_t const* data_ptr;
-    size_type size;
-    std::tie(data_ptr, size) = [&]() {
-      if (src.head()) {
-        auto const ptr = base_ptr + current_info->dst_offset;
-        // if we have data, num_elements will always be the correct size.
-        // we don't want to use num_rows because if we are an offset column, num_rows
-        // represents the # of rows of our owning parent. num_elements always represents
-        // the proper size for this column
-        auto const size = current_info->num_elements;
-        ++current_info;
-        return std::make_pair(ptr, size);
-      }
-      // Parent columns w/o data (e.g., strings, lists) don't have an associated `dst_buf_info`,
-      // therefore, use the first child's info if it has at least one child. Their num_rows value
-      // will be correct (also see comment above)
-      auto const size = (src.num_children() == 0) ? 0 : current_info->num_rows;
-      return std::make_pair(static_cast<uint8_t const*>(nullptr), size);
-    }();
-    auto children = std::vector<column_view>{};
-    children.reserve(src.num_children());
-    current_info = build_output_columns(
-      src.child_begin(), src.child_end(), current_info, std::back_inserter(children), base_ptr);
-    return column_view{src.type(), size, data_ptr, bitmask_ptr, null_count, 0, std::move(children)};
-  });
+  std::transform(
+    begin, end, out_begin, [&current_info, base_ptr, &metadata](column_view const& src) {
+      // Use C++17 structured bindings
+      bitmask_type const* bitmask_ptr;
+      size_type null_count;
+      std::tie(bitmask_ptr, null_count) = [&]() {
+        if (src.nullable()) {
+          auto const ptr =
+            reinterpret_cast<bitmask_type const*>(base_ptr + current_info->dst_offset);
+          ++current_info;
+          return std::make_pair(ptr, UNKNOWN_NULL_COUNT);
+        }
+        return std::make_pair(static_cast<bitmask_type const*>(nullptr), 0);
+      }();
+
+      // size/data pointer for the column
+      auto const size = current_info->num_elements;
+      uint8_t const* data_ptr =
+        size == 0 || src.head() == nullptr ? nullptr : base_ptr + current_info->dst_offset;
+      ++current_info;
+
+      // children
+      auto children = std::vector<column_view>{};
+      children.reserve(src.num_children());
+
+      // add metadata
+      add_column_metadata(
```

Comment:
I get that it'd probably be tricky to make the metadata a return from the enclosing function, but it seems like this function would be easy enough to return instead of pass by ref. In fact, it may be easier all together to just use an IILE instead of a function: